design-docs/001-virtual-chunk-improvements.md

```diff
@@ -0,0 +1,259 @@
+
+```rust
+// this is pseudo-code
+struct Config {
+  inline_chunk_threshold_bytes: u16,
+  ... more fields
+
+  virtual_chunk_containers: Vec<VirtualChunkContainer>,
+}
+```
+
+The `Config` object acts as a default, overridable by users at runtime. But different fields of configuration may have different behaviors on overrides.
+
+### New mechanism to set virtual references
+
+We add an optional `checksum` argument to `set_virtual_ref` (we need a better name for this argument).
+
+```python
+# this is pseudo-code
+
+class ETag:
+  value: str
+
+class LastModified:
+  value: datetime
+
+class Repository:  # the class with this method is going to change soon to be Session, for an unrelated reason
+  def set_virtual_ref(
+      self,
+      key: str,
+      location: str,
+      offset: int,
+      length: int,
+      checksum: ETag | LastModified | None = None,
+      validate_containers: bool = False
+  ) -> None: ...
+```
+
+`checksum` can be set to the in-object-store ETag of the file containing the virtual chunk, or to its last modified time. In the case of last modified time, the value
+doesn't need to be precise. The system will only fail to provide the chunk if the object's last-modified-at is larger than the passed value. So, passing `checksum = LastModified(datetime.now())` will usually be enough.
+
+This new argument will help us satisfy requirement 4.
+
+The interpretation of the `location` argument also changes. `location` must be an absolute url that includes a protocol. The protocol can be:
+
+* s3://
+* gcs://
+* tigris://  (is this the right protocol)
+* fs://  (is this right for local file system)
+* etc. for other supported object stores
+* arbitrary-virtual-chunk-container-name://  (example: models://)
+
+Depending on the protocol used, Icechunk will use a different mechanism to resolve the chunk, as explained below.
+
+If `validate_containers` is set to `True`, the function will fail if the protocol passed is not registered in a virtual chunk container. Using `validate_containers=False` is useful, if containers will be declared after the write.
+
+### Reading virtual chunks
+
+When a virtual chunk is requested, Icechunk will find in the `location` field of the stored reference, an absolute URL with a specific protocol. It will need to map this location to a given object store, bucket, path and set of credentials.
+
+The mapping is done by comparing the protocol and path to the set of configured virtual chunk containers, plus a set of default virtual chunk containers.
+
+This is how the mapping from location to cloud+bucket+prefix+credentials works:
+
+* Using the `location` protocol we find all containers that match the protocol
+* We sort the matching containers in order of descending prefix length
+* We find the first one that has a `bucket_prefix` that is a prefix of the `location` path (this will be the most specific container that matches)
```

Comment:
What do buckets have anything to do with here?
Prefix is a string of any length, no?

Reply:
The chunk's URL is `s3://foo/bar/baz.nc`, we need to find a container that matches it to know what set of credentials and object store client configuration to use. If there are two containers that work for protocol s3, the one with `bucket_prefix = foo/bar` will match, instead of the one with `bucket_prefix = hello` or the one with `bucket_prefix = foo`.

Reply:
"We find the container with the longest prefix matching the given URL"